src/AspNetCore/Src/applicationmanager.cxx

```diff
@@ -197,72 +222,194 @@ APPLICATION_MANAGER::RecycleApplication(
     HRESULT          hr = S_OK;
     APPLICATION_INFO_KEY  key;
     DWORD            dwPreviousCounter = 0;
+    APPLICATION_INFO_HASH* table = NULL;
+    CONFIG_CHANGE_CONTEXT context;
 
     hr = key.Initialize(pszApplicationId);
     if (FAILED(hr))
     {
         goto Finished;
     }
+
+    table = new APPLICATION_INFO_HASH();
+
+    if(table == NULL)
+    {
+        hr = E_OUTOFMEMORY;
+        goto Finished;
+    }
+
+    //
+    // few application expected, small bucket size for hash table
+    //
+    if (FAILED(hr = table->Initialize(17 /*prime*/)))
+    {
+        goto Finished;
+    }
+
+    context.pstrPath = pszApplicationId;
+
     AcquireSRWLockExclusive(&m_srwLock);
     dwPreviousCounter = m_pApplicationInfoHash->Count();
 
-    m_pApplicationInfoHash->DeleteKey(&key);
+    // We don't want to hold the lock for long time as it will block all incoming requests
+    // Make a shallow copy of existing hashtable as we may nee to remove nodes
+    // This also make sure application shutdown will not be called inside the lock
+    m_pApplicationInfoHash->Apply(APPLICATION_INFO_HASH::ReferenceCopyToTable, static_cast<PVOID>(table));
+    DBG_ASSERT(dwPreviousCounter == table->Count());
+    
+    // Removed the applications which are impacted by the configurtion change
+    m_pApplicationInfoHash->DeleteIf(FindConfigChangedApplication, (PVOID)&context);
 
-    if(dwPreviousCounter != m_pApplicationInfoHash->Count())
+    if (dwPreviousCounter != m_pApplicationInfoHash->Count())
     {
-        // Application got recycled. Log an event
-        STACK_STRU(strEventMsg, 256);
-        if (SUCCEEDED(strEventMsg.SafeSnwprintf(
-            ASPNETCORE_EVENT_RECYCLE_CONFIGURATION_MSG,
-            pszApplicationId)))
+        if (m_hostingModel == HOSTING_IN_PROCESS)
         {
-            UTILITY::LogEvent(g_hEventLog,
-                EVENTLOG_INFORMATION_TYPE,
-                ASPNETCORE_EVENT_RECYCLE_CONFIGURATION,
-                strEventMsg.QueryStr());
+            APPLICATION_INFO_HASH* tmp = m_pApplicationInfoHash;
+            m_pApplicationInfoHash = table;
+            table = tmp;
+            // Keep the original applicationinfo hashtable to continue serve request
+            // Trigger a worker process recycle and let the process shutdown code path to handle it
+            // So that we will drop/reject the incoming requests before WAS spins another worker process
+            g_pHttpServer->RecycleProcess(L"AspNetCore Recycle Process on Demand Due to In-process Application Configuration Changed");
         }
-
     }
 
     if (m_pApplicationInfoHash->Count() == 0)
     {
         m_hostingModel = HOSTING_UNKNOWN;
     }
 
+    ReleaseSRWLockExclusive(&m_srwLock);
 
-    if (g_fAspnetcoreRHLoadedError)
+    if(!context.MultiSz.IsEmpty() && (m_hostingModel == HOSTING_OUT_PROCESS || m_hostingModel == HOSTING_UNKNOWN))
```

Comment:
For out-proc, we do update m_pApplicationInfoHash within the lock.
What would be the behavior of the requests coming to the app right after the table update?
What would be the behavior of the requests being processed in the backend app? Within shutdown limit, if their responses come, does ANCM still forward them back to clients?